### PR TITLE
feat(kurtosis): Forge script to deploy smart contracts

### PR DESCRIPTION
**Pull request by nidhi-singh02**
In the first iteration, displaying the content of the file where transaction is saved.

## Summary by CodeRabbit

- **New Features**
	- Introduced a structured approach for managing the Kurtosis environment, enhancing local development and testing for decentralized applications.
	- Added a configuration template for deploying smart contracts, streamlining the setup process for developers within the Berachain ecosystem.
	- Implemented a robust framework for automated smart contract deployments using the Forge toolset, significantly improving deployment efficiency.

- **Documentation**
	- Added comprehensive README instructions for deploying smart contracts using the Kurtosis framework, detailing configuration and operational guidelines.

### CODE REVIEW

**Comment by nidhi-singh02** [comment]
Currently displaying the full file content where transactions are saved, we could tweak around with what needs to be displayed to give better developer experience.

**Comment by nidhi-singh02** [comment]
Currently displaying the full file content where transactions are saved, we could tweak around with what needs to be displayed to give better developer experience.

File: testing/forge-script/main.star [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    transaction_file = plan.exec(
+        service_name = foundry_service.name,
+        recipe = ExecRecipe(
+            command = ["/bin/sh", "-c", "cd /app/contracts && {} ".format(transaction_file)],
+        ),
+    )
+    exec_output = plan.exec(
+        service_name = foundry_service.name,
+        recipe = ExecRecipe(
+            command = ["/bin/sh", "-c", "cat {}".format(transaction_file["output"])],
```
Comment (nidhi-singh02) [inline testing/forge-script/main.star:55]: Currently displaying the full file content where transactions are saved, we could tweak around with what needs to be displayed to give better developer experience.

File: testing/forge-script/main.star [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+            },
+        ),
+    )
+
+    plan.exec(
+        service_name = foundry_service.name,
+        recipe = ExecRecipe(
+            command = ["/bin/sh", "-c", "cd /app/contracts && forge build"],
+        ),
+    )
```
[the remark this answers is not in the document]
Comment (nidhi-singh02) [reply testing/forge-script/main.star]: with above snippet, my kurtosis package is not built get error Forge build failed all the time

File: testing/forge-script/main.star [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    deploy_contracts(plan, deployment)
+
+# Define the function to run the Forge script for deployment
+def deploy_contracts(plan, deployment):
+    contract_name = deployment["contract_name"]
+    script_path = deployment["script_path"]
+    repository = deployment["repository"]
+    rpc_url = deployment["rpc_url"]
+    private_key = deployment["private_key"]
+
```
[the remark this answers is not in the document]
Comment (nidhi-singh02) [reply testing/forge-script/main.star:15]: What is the best way to handle private key

File: testing/forge-script/forge-config.yaml [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, NON-INFRINGEMENT, AND
+# TITLE.
+
+deployment:
+  repository: "github.com/nidhi-singh02/solidity-scripting/"    # follow the same format as defined here.
+  script_path: "script/NFT.s.sol"  # this must be relative to the repository path
+  contract_name: "MyScript"
+  rpc_url: "https://bartio.rpc.berachain.com/"
+  private_key: "0x0"
+# TODO: Support all wallet options later on
```
[the remark this answers is not in the document]
Comment (nidhi-singh02) [reply testing/forge-script/forge-config.yaml:27]: support all wallet options for me.

File: testing/forge-script/forge-config.yaml [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+# LICENSOR AS EXPRESSLY REQUIRED BY THIS LICENSE).
+#
+# TO THE EXTENT PERMITTED BY APPLICABLE LAW, THE LICENSED WORK IS PROVIDED ON
+# AN “AS IS” BASIS. LICENSOR HEREBY DISCLAIMS ALL WARRANTIES AND CONDITIONS,
+# EXPRESS OR IMPLIED, INCLUDING (WITHOUT LIMITATION) WARRANTIES OF
+# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, NON-INFRINGEMENT, AND
+# TITLE.
+
+deployment:
+  repository: "github.com/nidhi-singh02/solidity-scripting/"  # follow the same format as defined here.
```
Comment (itsdevbear) [inline testing/forge-script/forge-config.yaml:22]: don't do this

File: testing/forge-script/forge-config.yaml [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+  script_path: "script/NFT.s.sol"  # this must be relative to the repository path
+  contract_name: "MyScript"
+  rpc_url: "https://bartio.rpc.berachain.com/"
```
Comment (itsdevbear) [inline testing/forge-script/forge-config.yaml:25]: make default integration with the localnet

Comment (nidhi-singh02) [reply testing/forge-script/forge-config.yaml:25]: Makes sense. Let me change.

File: testing/forge-script/forge-config.yaml [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, NON-INFRINGEMENT, AND
+# TITLE.
+
+deployment:
+  repository: "github.com/nidhi-singh02/solidity-scripting/"  # follow the same format as defined here.
+  script_path: "script/NFT.s.sol"  # this must be relative to the repository path
+  contract_name: "MyScript"
+  rpc_url: "http://localhost:8551"
+  wallet:
+    type: "private_key"  # currently only private_key wallet is supported. Do not change the type.
```
[the remark this answers is not in the document]
Comment (nidhi-singh02) [reply testing/forge-script/forge-config.yaml:27]: nope.

File: testing/forge-script/forge-config.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (nidhi-singh02) [reply testing/forge-script/forge-config.yaml:22]: It is same only.

Comment (nidhi-singh02) [reply testing/forge-script/forge-config.yaml:22]: I tried script path approach earlier, as deployment script has dependency on the source code in src folder, forge script fails. Hence uploading the whole contracts repository.

Comment (nidhi-singh02) [reply testing/forge-script/forge-config.yaml:22]: There should be two ways - github and local files/folders.

File: testing/forge-script/forge-config.yaml [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+#
+# TO THE EXTENT PERMITTED BY APPLICABLE LAW, THE LICENSED WORK IS PROVIDED ON
+# AN “AS IS” BASIS. LICENSOR HEREBY DISCLAIMS ALL WARRANTIES AND CONDITIONS,
+# EXPRESS OR IMPLIED, INCLUDING (WITHOUT LIMITATION) WARRANTIES OF
+# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, NON-INFRINGEMENT, AND
+# TITLE.
+
+deployment:
+  repository: "github.com/nidhi-singh02/solidity-scripting/"  # follow the same format as defined here.
+  script_path: "script/NFT.s.sol"  # this must be relative to the repository path
```
Comment (ocnc) [inline testing/forge-script/forge-config.yaml:23]: why not just use the regular DeployAndCallERC20.s.sol already in `/contracts`

Comment (nidhi-singh02) [reply testing/forge-script/forge-config.yaml:23]: agreed, will make it work for both local contracts as well as contracts hosted on github. Will test DeployAndCallERC20.s.sol.

File: testing/forge-script/forge-config.yaml [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+  contract_name: "MyScript"
+  rpc_url:
+    host_ip: "Host IP address"  # <Host IP address>
+    port: 8547  # Spin up local devnet via kurtosis, then public port is 8547
+  wallet:
```
[the remark this answers is not in the document]
Comment (nidhi-singh02) [reply testing/forge-script/forge-config.yaml:28]: where are the trailing spaces, it is correct I think.

File: testing/forge-script/forge-config.yaml [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+# AN “AS IS” BASIS. LICENSOR HEREBY DISCLAIMS ALL WARRANTIES AND CONDITIONS,
+# EXPRESS OR IMPLIED, INCLUDING (WITHOUT LIMITATION) WARRANTIES OF
+# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, NON-INFRINGEMENT, AND
+# TITLE.
+
+deployment:
+  repository: "github.com/nidhi-singh02/solidity-scripting/"  # follow the same format as defined here.
+  script_path: "script/NFT.s.sol"  # this must be relative to the repository path
+  contract_name: "MyScript"
+  rpc_url:
```
Comment (nidhi-singh02) [inline testing/forge-script/forge-config.yaml:25]: Will change this to just accept rpc_url so it works with all the cases.

File: testing/forge-script/main.star [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+            rpc_url,
+            wallet_command,
+        ),
+    )
+
+    # Get the forge script output in a output.json file and grep from it
+    transaction_file = "grep 'Transactions saved to' output.json | awk -F': ' '{print $2}'"
+    plan.print("transaction_file", transaction_file)
+
+    transaction_file_details = exec_on_service(plan, foundry_service.name, "cd {} && {}".format(contract_path, transaction_file))
```
[the remark this answers is not in the document]
Comment (nidhi-singh02) [reply testing/forge-script/main.star:65]: it is same right.

File: testing/forge-script/forge-config.yaml [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+#
+# TO THE EXTENT PERMITTED BY APPLICABLE LAW, THE LICENSED WORK IS PROVIDED ON
+# AN “AS IS” BASIS. LICENSOR HEREBY DISCLAIMS ALL WARRANTIES AND CONDITIONS,
+# EXPRESS OR IMPLIED, INCLUDING (WITHOUT LIMITATION) WARRANTIES OF
+# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, NON-INFRINGEMENT, AND
+# TITLE.
+
+# Please refer to the README.md file for more information on how to fill this file.
+deployment:
+  repository: "github.com/nidhi-singh02/beacon-kit"  # give repo name if there are submodules, else give the folder till contracts
```
[the remark this answers is not in the document]
Comment (nidhi-singh02) [reply testing/forge-script/forge-config.yaml:23]: seems as per the suggestion, I see same spacing.

File: testing/forge-script/main.star (hunk 11 above)
[the remark this answers is not in the document]
Comment (nidhi-singh02) [reply testing/forge-script/main.star:65]: it is same I see

File: testing/forge-script/forge-config.yaml (hunk 8 above)
Comment (nidhi-singh02) [reply testing/forge-script/forge-config.yaml]: For local files/folders, there isn't a direct way to access them via kurtosis unless they are part of the package itself. Tested ` DeployAndCallERC20.s.so` through Github url (forked beacon-kit under my profile). It works fine.

File: testing/forge-script/README.md [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+
+- Use github URL.
+
+    Edge scenario: 
+    This would not be supported if the contracts are part of this repository - beacon-kit, in that case, fork the repoistory into your user profile and use that as a repository.
+
+    I know this is a kind of workaround, until Kurtosis supports local URLs, we don't really have a choice.
+
+    **Example for running contracts in beacon-kit**
+    ```bash
```
[the remark this answers is not in the document]
Comment (nidhi-singh02) [reply testing/forge-script/README.md:43]: this is not part of the bash script

**Review by corduroybera — APPROVED** [review]

**Pull request merged**